std.algorithms: document public methods

wilzbach opened the pull request:
Here's a small project that I intend to do when I am on the road - document the public methods in Phobos or set them private etc ;-)
As you can imagine, many parts in Phobos aren't properly ddoced - so let's do this bit by bit.

If someone wants to help out, here's the [current output of dscanner](http://sprunge.us/LjXd).

There are some fixes which might be controversial - I will mark them so.

wilzbach commented on std/algorithm/searching.d:
```diff
@@ -1786,9 +1789,7 @@ if (isRandomAccessRange!R1 && isBidirectionalRange!R2
-        // @@@BUG@@@
-        //return haystack[$ .. $];
-        return haystack[haystack.length .. haystack.length];
+        return haystack[$ .. $];
```
it seems to me that this bug is fixed?

wilzbach commented on std/algorithm/searching.d:
```diff
@@ -1938,7 +1938,7 @@ if (isRandomAccessRange!R1 && isForwardRange!R2 && !isBidirectionalRange!R2 &&
 // Internally used by some find() overloads above. Can't make it
 // private due to bugs in the compiler.
-/*private*/ R1 simpleMindedFind(alias pred, R1, R2)(R1 haystack, R2 needle)
+private R1 simpleMindedFind(alias pred, R1, R2)(R1 haystack, R2 needle)
```
it seems that this bug is fixed?

JackStouffer commented:
@burner IIRC you're the one who made all of the `alias string this` overloads, any reason they shouldn't be documented?

burner commented:
@JackStouffer I don't recall this. Anyway, if they are public functions and are supposed to be public then the should be documented.

burner commented:
this PR is mixing documentation with general maintenance and visibility changes. Could you split that into multiple PRs. That would make reviewing easier and thereby make pulling or closing faster.

JackStouffer commented on std/algorithm/iteration.d:
```diff
@@ -1319,7 +1319,7 @@ private struct FilterBidiResult(alias pred, Range)
 
-// group
+/// group
 struct Group(alias pred, R) if (isInputRange!R)
```
This should be under the same docs as its factory function. Same goes for all of the public range structs.

wilzbach replied:
so `/// ditto`?

JackStouffer replied:
Yeah, and you should move it to under the factory function. No one should be using the struct directly, but since it's public we can't make it a static struct without deprecation cycles that are of questional value anyway.

wilzbach commented:
> this PR is mixing documentation with general maintenance and visibility changes. Could you split that into multiple PRs. That would make reviewing easier and thereby make pulling or closing faster.

@burner so that I understand right - setting accidentally exposed symbols to `private` should be a separate PR?

burner commented:
@wilzbach yes, because it is a bugfix and not a documentation change. As said, this will speed the whole process. The little extra time you will spend branching you will make back double/tripple during review.

JackStouffer commented on std/algorithm/iteration.d:
```diff
@@ -4874,13 +4880,14 @@ private struct UniqResult(alias pred, Range)
 {
-    size_t[] indices, state;
-    Range r;
+    private size_t[] indices, state;
```
You need to deprecate these, perferably in a different PR. See https://github.com/dlang/phobos/pull/3950/files

wilzbach commented:
> @wilzbach yes, because it is a bugfix and not a documentation change. As said, this will speed the whole process. The little extra time you will spend branching you will make back double/tripple during review.

done - this PR now only contains documentation changes and as @JackStouffer suggested I moved undocumented `struct`'s below the method and added the magic `/// ditto`.

JackStouffer commented:
LGTM

wilzbach commented:
> LGTM

ping

schveiguy commented:
Based on the recent discussions on Voldemort types, I wonder, shouldn't these actually be Voldemort? I wouldn't make them so, because this can bloat the binary incredibly fast. But I don't know if we should expose the types as public API.

I'm working on a possible solution to the bloat that would allow us to have the types as Voldemort but not with the bloat.

burner commented:
LGTM

burner commented:
Auto-merge toggled on

wilzbach commented:
> But I don't know if we should expose the types as public API.

This just adds them to the documentation output - they are already publicly exposed and would need a deprecation cycle if we make them private/Voldemort again,.

schveiguy commented:
(as an aside, github is really borked for me today, it doesn't show all the discussion).

It's a much easier deprecation to remove undocumented public symbols than it is to remove public API.

schveiguy commented:
In any case, we have until the next major release to decide.

burner commented:
I did just turn auto merge on. but github does not display it. Lets see what explodes and then take action
